sycl/source/detail/program_manager/program_manager.hpp

```diff
@@ -359,8 +360,16 @@ class ProgramManager {
   ~ProgramManager() = default;
 
   template <typename NameT>
-  bool kernelUsesAssert(const NameT &KernelName) const {
-    return m_KernelUsesAssert.find(KernelName) != m_KernelUsesAssert.end();
+  bool kernelUsesAssert(const NameT &KernelName,
+                        KernelNameBasedCacheT *KernelNameBasedCachePtr) const {
+    if (!KernelNameBasedCachePtr)
+      return m_KernelUsesAssert.find(KernelName) != m_KernelUsesAssert.end();
+
+    std::optional<bool> &UsesAssert = KernelNameBasedCachePtr->UsesAssert;
+    if (!UsesAssert.has_value())
+      UsesAssert =
```

Comment:
Should we assign `UsesAssert` in the same place where we insert in the `m_KernelUsesAssert`?

Reply:
`m_KernelUsesAssert` is populated during device image registration on application startup (or while loading a library), at which point we don't have access to any of the header based caches.

Reply:
As I suggested in the #18654, can we init the `KernelNameBasedCachePtr->UsesAssert` when an instance of the `KernelNameBasedCacheT` is created?